File: spring_boot/Microservices/country-service/src/main/java/com/manipal/controller/CountryController.java
package com.manipal.controller;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import com.manipal.demo.model.Capital;
import com.manipal.demo.model.Country;
import com.manipal.demo.model.CountryCodes;
import com.manipal.demo.proxy.CapitalServiceProxy;
import com.manipal.demo.repository.CountryCodesRepository;
import com.manipal.demo.repository.CountryRepository;

@RestController
public class CountryController {
	@Autowired
	private CapitalServiceProxy proxy;
	
	@Autowired
	CountryRepository repo;
	
	@Autowired
	CountryCodesRepository countryCodesRepo;
	
	@GetMapping("/home")
	public String home() {
		return "Country-service";
	}
	@PostMapping("country-service/add/codes")
	public CountryCodes addCodes(@RequestBody CountryCodes countryCodes) {
		return countryCodesRepo.save(countryCodes);
	}
	
	@GetMapping("/country-service/country/{countryName}")
	public Country getCountryDetails(@PathVariable String countryName) {
		CountryCodes countryCodes= countryCodesRepo.findByCountryName(countryName);
		System.out.println("\n\n======\n"+countryCodes);
		Capital capital = proxy.retriveCapital(countryCodes.getCountryCode());
		Country country = new Country();
		country.setCountryName(countryName);
		country.setCountryCode(countryCodes.getCountryCode());
		country.setCapital(capital);
		System.out.println("\n\n\n------\n"+country+"-----\n\n\n\n");
		//repo.save(country);
		return country;			
		
	}
	@PostMapping("/country-service/country/{countryName}")
	public Country putCountryDetails(@PathVariable String countryName) {
		CountryCodes countryCodes= countryCodesRepo.findByCountryName(countryName);
		System.out.println("\n\n======\n"+countryCodes);
		Capital capital = proxy.retriveCapital(countryCodes.getCountryCode());
		Country country = new Country();
		country.setCountryName(countryName);
		country.setCountryCode(countryCodes.getCountryCode());
		country.setCapital(capital);
		System.out.println("\n\n\n------\n"+country+"-----\n\n\n\n");
		repo.save(country);
		return country;		
		
	}
	
	@GetMapping("/country-service/show-all-codes")
	public List<CountryCodes> getAll(){
		return countryCodesRepo.findAll();
	}
	@GetMapping("/country-service/all")
	public List<Country> getAllCountries(){
		return repo.findAll();
	}
	
	

}

File: spring_boot/bootKotlin/src/main/kotlin/com/sourabh/bootKotlin/controller/CountryController.kt
package com.sourabh.bootKotlin.controller

import com.sourabh.bootKotlin.model.Country
import com.sourabh.bootKotlin.service.CountryService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
class CountryController(@Autowired private val countryService: CountryService) {

    @GetMapping("/home")
    fun home():String{
        return "Hello Mortals!!"
    }
    @PostMapping("/countries")
    fun addOrUpdate(@RequestBody country: Country):Country{
        return countryService.addOrUpdateCountry(country)
    }
    @PostMapping("/countries/bulk")
    fun addBulk(@RequestBody countries: MutableList<Country>): MutableList<Country> {
        return countryService.addBulk(countries)
    }
    @GetMapping("/countries")
    fun getAll(): MutableList<Country?> {
        return countryService.getAll()
    }
    @GetMapping("/countries/{id}")
    fun getCountry(@PathVariable id: Int): Country? {/*
        val country = countryService.getCountry(id)
        val linkBuilder= linkTo(methodOn(CountryController.class).getAll())
        val linkBuilder1: WebMvcLinkBuilder = WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.methodOn().retrieveTodo(id))
        val entityModel: EntityModel<*> = EntityModel.of<Any>(toDo)
        entityModel.add(linkBuilder.withRel("all-todos"))
        entityModel.add(linkBuilder1.withSelfRel())*/
        return countryService.getCountry(id)
    }
    @GetMapping("/countries/name/{name}")
    fun getCountryByName(@PathVariable name: String):Country?{
        return countryService.getCountryByName(name)
    }
    @GetMapping("/countries/lang/{lang}")
    fun getEnglishCountries(@PathVariable lang: String): MutableList<Country?>? {
        return countryService.getContriesLang(lang)
    }
    @GetMapping("/countries/lt/{value}")
    fun countryPopulationLessThan(@PathVariable value: Int):MutableList<Country?>{
        return countryService.countryPopulationLessThan(value)
    }

    @PutMapping("/countries/{id}")
    fun updateCountry(@PathVariable id: Int, @RequestBody countryDetails: Country): ResponseEntity<Any?> {
        val countryOld = countryService.getCountry(id)
        countryOld!!.countryId=countryDetails.countryId
        countryOld.name =countryDetails.name
        countryOld.capital=countryDetails.capital
        countryOld.language=countryDetails.language
        countryOld.population=countryDetails.population
        val countryUpdated= countryService.addOrUpdateCountry(countryOld)
        val headers = HttpHeaders()
        headers.add("Response-from", "CountryController")
        return ResponseEntity<Any?>(countryUpdated, headers, HttpStatus.OK)
    }
    @DeleteMapping("/countries/{id}")
    fun deleteCountry(@PathVariable id: Int): String {
        val country = countryService.getCountry(id)
        countryService.deleteCountry(id)
        return "Deleted record for country id: ${country?.countryId} name: ${country?.name}"

    }

}